File: DanglingNullPointer/DanglingNullPointer.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    int *ptr = (int *)malloc(sizeof(int)); // Allocate memory
    if (ptr == NULL) {
        printf("Memory allocation failed\n");
        return 1;
    }

    *ptr = 42; // Assign a value to the allocated memory
    printf("Value: %d\n", *ptr);

    free(ptr); // Free the allocated memory
    // ptr = NULL; // Set pointer to NULL to avoid dangling pointer

    // Attempting to access the memory after it has been freed
    if (ptr != NULL) {
        printf("Value after free: %d\n", *ptr); // This line will not execute
    } else {
        printf("Pointer is NULL, cannot dereference\n");
    }

    return 0;
}

File: DanglingNullPointer/DanglingNullPointer.rs
fn main() {
    // Allocate memory for an integer on the heap using Box
    let mut ptr = Box::new(100);

    // Use the allocated memory
    println!("Value: {}", *ptr);

    // After this block, 'ptr' is automatically dropped, 
    // and the memory is freed safely, preventing dangling pointer issues.

    // Manually dropping ptr early (optional)
    std::mem::drop(ptr); // This explicitly frees the memory early

    // After dropping, 'ptr' cannot be used anymore, preventing dangling pointer access
    // Trying to use 'ptr' now would result in a compile-time error.
    println!("Value: {}", *ptr); // Uncommenting this line would cause a compile-time error
    // *ptr = 200; // Uncommenting this line would cause a compile-time error
}